### DISCUSSION TITLE
Can I maintain internal state inside a StateNotifier

### DISCUSSION THREAD

**Discussion by otto-dev** [rrousselGit/riverpod#1963] [Q&A] [answered]
Can I maintain internal, mutable state inside a `StateNotifier`, such as `Timers`, if the  `StateNotifierProvider` does not `watch` (only `read`) other providers?

I'm talking about internal state which is separate from the `state` class field, and hence is not exposed to the `StateNotifierProvider`. 

An example would be a class field `Timer? _debounce;`, which I use to debounce a certain action, such as fetching data from the backend for the `state` field.

**Comment by rrousselGit** [comment]
Not with StateNotifier

With Notifier though, yes.

**Comment by TimWhiting** [comment, marked as answer]
Technically yes, (I've done it before, and you could try it out). As long as you don't need the UI to know about the internal state, and you update the user visible state when something interesting happens with the internal state. Make sure to dispose the internal state in the dispose method.

I would recommend like Remi says to use Notifier though. It seems to be the way of the future.

**Reply by rrousselGit** [reply]
Ah I got confused by the first line. I thought it was about preserving the StateNotifier properties if the provider _did_ watch another provider

My answer was referring to how Notifier isn't recreated on `watch` change. The instance is preserved, and dependency change simply re-execute `Notifier.build`.
So you can preserve state across watch changes

**Reply by otto-dev** [reply]
Thanks @TimWhiting for this comprehensive answer. Appreciate it!

> `Notifier` isn't recreated on `watch` change. The instance is preserved, and dependency change simply re-execute `Notifier.build`.

That's really good info. I'll stick with `Notifier` then.

Thanks for all the great work you put into this library!
